pkg/workflows/wasm/workflowLib/plugin.go

```diff
@@ -0,0 +1,270 @@
+		Requests:  requests,
+		Finished:  unscheduledFinishedExecutionIDs,
+	}
+
+	return proto.Marshal(observation)
+}
+
+func (p *workflowLibPlugin) ValidateObservation(ctx context.Context, oc ocr3types.OutcomeContext, _ types.Query, ao types.AttributedObservation) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	observation := &pb.Observation{}
+	if err := proto.Unmarshal(ao.Observation, observation); err != nil {
+		return err
+	}
+
+	priorOutcome := &pb.Outcome{}
+	if err := proto.Unmarshal(oc.PreviousOutcome, priorOutcome); err != nil {
+		return err
+	}
+	if priorOutcome.ObservedDonTimes == nil {
+		priorOutcome.ObservedDonTimes = map[string]*pb.ObservedDonTimes{}
+	}
+
+	newInvalidRequestError := func(requestSeqNum int64, id string, currSeqNum int) error {
+		return fmt.Errorf("request number %d for id %s is greater than the number of observed don times %d", requestSeqNum, id, currSeqNum)
+	}
+
+	// A DON time beyond the expected number of requests is invalid, as there was never consensus on the prior request, which should be blocking.
+	var err error
+	for id, requestedSeqNumber := range observation.Requests {
+		times, ok := priorOutcome.ObservedDonTimes[id]
+		if !ok {
+			if requestedSeqNumber != 0 {
+				err = errors.Join(err, newInvalidRequestError(requestedSeqNumber, id, 0))
+			}
+			continue
+		}
+		seqNum := len(times.Timestamps)
+		if requestedSeqNumber > int64(seqNum) {
+			err = errors.Join(err, newInvalidRequestError(requestedSeqNumber, id, seqNum))
+		}
+	}
+
+	return err
+}
+
+func (p *workflowLibPlugin) ObservationQuorum(_ context.Context, _ ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (quorumReached bool, err error) {
+	return quorumhelper.ObservationCountReachesObservationQuorum(quorumhelper.QuorumTwoFPlusOne, p.config.N, p.config.F, aos), nil
+}
+
+func (p *workflowLibPlugin) Outcome(ctx context.Context, outctx ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (ocr3types.Outcome, error) {
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+
+	NumFinishedRequests := map[string]int64{} // counts how many nodes reported where a new DON timestamp might be needed
+	finishedNodes := map[string]int64{}       // counts number of nodes finished with the workflow for executionID
+	var times []int64
+
+	for _, ao := range aos {
+		observation := &pb.Observation{}
+		if err := proto.Unmarshal(ao.Observation, observation); err != nil {
+			return nil, err
+		}
+
+		prevOutcome := &pb.Outcome{}
+		if err := proto.Unmarshal(outctx.PreviousOutcome, prevOutcome); err != nil {
+			return nil, err
+		}
+
+		for id, currSeqNum := range observation.Requests {
+			if err := ctx.Err(); err != nil {
+				return nil, err
+			}
+			if _, ok := prevOutcome.ObservedDonTimes[id]; !ok {
+				prevOutcome.ObservedDonTimes[id] = &pb.ObservedDonTimes{}
+			}
+			if currSeqNum == int64(len(prevOutcome.ObservedDonTimes[id].Timestamps)) {
+				NumFinishedRequests[id]++
```

Comment:
It took me a very long time to understand this logic :) Please add a comment explaining the we only count requests for the next sequence number and ignore all other ones.